pkg/sql/sem/builtins/builtins.go

```diff
@@ -6223,6 +6223,33 @@ table's zone configuration this will return NULL.`,
 			Volatility: tree.VolatilityVolatile,
 		},
 	),
+
+	"crdb_internal.set_job_debug_pausepoint": makeBuiltin(
```

Comment:
I'm still confused... I recall that early on I had a similar change introduced
where you could register "test value" points -- named things that can invoke
a function, or change or update some value, etc... The problem of course
is that this ease (i.e. adding arbitrary name breakpoints) was the reason why it
didn't fly -- use TestingKnobs was the answer (for better or worse)...

Now, this function only operates on a single node registry.  So, it appears that the only
use case is in single node deployment -- i.e. tests, and not even all of them.
So, those registry methods could be made into a test only implementation...  And then I
don't see how this crdb_internal would be useful...  Also, consider an abuse verctor where we
create lots of breakpoints, with long names, and never clean them up.  That's not great.

Do you have a concrete production code/real binary use case you have in mind?

Reply:
> Now, this function only operates on a single node registry. So, it appears that the only
use case is in single node deployment -- i.e. tests, and not even all of them. So, those registry methods could be made into a test only implementation.

I don't think that's entirely true -- you could use this in a roachtest by just iterating over all three nodes to set it. You can use it in demo, or start-single-node, which you just can't do with testing knobs, and it is generally _difficult_ to use testingKnobs in logictests (you can go extend the logictest itself to have some directives that then enable the knobs, but it's not pleasant).

> an abuse verctor where we create lots of breakpoints
 Ah, I forgot to make the builtin admin-only.

> Do you have a concrete production code/real binary use case you have in mind?
Several. For starters, I want to start writing sql logictests that can do set_pausepoint and then pb_to_json to inspect job state mid job, writing a couple lines of SQL to inspect mid-job state instead of hundreds of lines of go testing knob, testserver, etc.

I also want to be able to do more one-off digging in `demo` or `start-single-node`.

I want to be able to ask roachtests to pause jobs at specific points too, where we also cannot use testing knobs today, e.g. to ensure a backup can be run _mid_ schema change or something, without just hoping we pick something that takes just about long enough we catch it in the right phase.

And then finally, we had a couple real production cases where it'd have been really useful to pause a job right before some step where it'd get into trouble and then rollback, such as a RESTORE that was failing right at the end. If we'd had a pausepoint close enough in that case, we could've use it to run until then, reliably paused and then done our manual intervention.

Reply:
We could make this _not_ per-node, and use a cluster setting (probably not one-per-point, but a comma-separated list of points like the env var) but cluster settings have some fixed costs if they're not used (use up a slot in mem even if you never set it) so it just didn't seem worth it for something _mostly_ used in tests, since I can just call the builtin on every node if I need to.